design-proposals/bound-token-for-vnc.md

```diff
@@ -0,0 +1,110 @@
+bound to a service account. They are mainly used in services running inside the cluster to authenticate
+between each other. 
+
+This document describes how to create a token that can be used to authenticate with the API server
+to only access VNC endpoint of a singe VM.
+
+## Motivation
+
+An administrator may want to give VNC access to a client application without giving it permissions to access
+anything else in the cluster. The VNC client could be pointed to VMI `/vnc` subresource and use
+the token to authenticate. 
+
+## Goals
+
+- A temporary token can be created to access VNC.
+- It can only be used with one specific VM.
+- It is valid only for a specific duration.
+
+## User Stories
+
+- A VM owner wants to give access to VNC to a third party application without giving it permissions to modify the VM.
+
+## Repos
+
+- https://github.com/kubevirt/kubevirt
+
+# Design
+
+An API subresource under `virtualmachineinstance` , `/vnc-token` is used to get a token for VNC.
+
+The token will refer to a `ServiceAccount` that only has permissions to access the `/vnc` subresource on
+one specific VM.
+
+
+## API Examples
+
+### Getting the VNC token
+```
+GET /apis/subresources.kubevirt.io/v1/namespaces/${VMI_NAMESPACE}/virtualmachineinstances/${VMI_NAME}/vnc-token
+```
+
+Parameters:
+- `duration` - Time duration while the token is valid.
+
+Response definition:
+```go
+type VncTokenResponse struct {
+	Token string `json:"token"`
+	ExpirationTimestamp metav1.Time `json:"expirationTimestamp"`
+}
+```
+
+### Accessing VNC using the token
+Access WebSocket endpoint:
+```
+/apis/subresources.kubevirt.io/v1/namespaces/${VMI_NAMESPACE}/virtualmachineinstances/${VMI_NAME}/vnc
+```
+with the following websocket protocols set:
+```
+base64url.bearer.authorization.k8s.io.<token>, base64.binary.k8s.io
+```
+
+## Virtctl examples
+
+### Getting the VNC token
+```bash
+virtctl vnc-token ${VMI_NAME}
+```
+
+### Accessing VNC using the token
+```bash
+# Login using the token
+oc login --token="${VNC_TOKEN}"
+
+# Access VNC
+virtctl vnc ${VMI_NAME}
+```
+
+## Scalability
+
+Creating a token requires an API call to the kubernetes API server.
+
+## Update/Rollback Compatibility
+
+### Update
+A new version of `virt-controller` will create and watch `ServiceAccounts`, `Roles` and `RoleBindings` for each VMI
+that has VNC enabled. 
+
+### Rollback
+The new resources will have `ownerReferences` pointing to the VMI, so even on rollback, they will be garbage collected
+when VMI is deleted.
+
+## Functional Testing Approach
+
+These cases should be tested:
+- Token can be created.
+- Token can be used to access VNC.
+- Token expires after a set time.
+- Token cannot be used to access VNC for a different VMI.
+- Token cannot be used to access any other resource.
+
+# Implementation Phases
+
+1. Add functionality to `virt-controller` to create `ServiceAccount`, `Role` and `RoleBinding` for every VMI.
```

Comment:
if two vmi are in same namespace, then also they need separate sa

Reply:
Correct, each SA would only have permissions to access one VMI.
But maybe we will implement a different design, I will update this document with alternatives today.